### PR TITLE
cmd/atlas/internal/cmdapi: make migrate apply logging template based

### PR DESCRIPTION
**Pull request by masseelch**
default
image

`--log '{{ json . "  " }}`
```json
{
  "Report": null,
  "Driver": "sqlite3",
  "URL": {
    "Scheme": "sqlite",
    "Opaque": "",
    "User": null,
    "Host": "",
    "Path": "",
    "RawPath": "",
    "OmitHost": false,
    "ForceQuery": false,
    "RawQuery": "mode=memory",
    "Fragment": "",
    "RawFragment": "",
    "DSN": "file:sqlite:?mode=memory",
    "Schema": "main"
  },
  "Dir": "../../build/migrations",
  "Pending": [
    {
      "Name": "1_changes.sql",
      "Version": "1",
      "Description": "changes"
    },
    {
      "Name": "2_asd.sql",
      "Version": "2",
      "Description": "asd"
    }
  ],
  "Applied": [
    {
      "Name": "1_changes.sql",
      "Version": "1",
      "Description": "changes",
      "Start": "2022-10-18T10:14:51.207977718+02:00",
      "End": "2022-10-18T10:14:51.208411998+02:00",
      "Stmts": [
        "CREATE TABLE `users` (`id` bigint NOT NULL, `name` varchar(255) NOT NULL, PRIMARY KEY (`id`));"
      ]
    },
    {
      "Name": "2_asd.sql",
      "Version": "2",
      "Description": "asd",
      "Start": "2022-10-18T10:14:51.208414373+02:00",
      "End": "2022-10-18T10:14:51.208533524+02:00",
      "Stmts": [
        "CREATE TABLE `groups` (`id` bigint NOT NULL, `name` varchar(255) NOT NULL, PRIMARY KEY (`id`));",
        "asdasd"
      ],
      "Error": {
        "SQL": "asdasd",
        "Error": "near \"asdasd\": syntax error"
      }
    }
  ],
  "Target": "2",
  "Start": "2022-10-18T10:14:51.207679841+02:00",
  "End": "2022-10-18T10:14:51.208533524+02:00",
  "Error": "sql/migrate: execute: executing statement \"asdasd\" from version \"2\": near \"asdasd\": syntax error"
}
```

### CODE REVIEW

File: cmd/atlas/internal/migrate/report.go [hunk 1]
```diff
@@ -297,6 +300,151 @@ func table(report *StatusReport) (string, error) {
...
+		a.End = f.End
+	}
+}
+
+// CountStmts returns the amount of applied statements.
+func (a *ApplyReport) CountStmts() (n int) {
+	for _, f := range a.Applied {
+		for range f.Stmts {
+			n++
+		}
```
Comment (a8m) [inline cmd/atlas/internal/migrate/report.go:416]:
```suggestion
		n += len(f.Stmts)
```

File: cmd/atlas/internal/migrate/report.go [hunk 2]
```diff
@@ -297,6 +300,151 @@ func table(report *StatusReport) (string, error) {
...
+		Error string `json:"Error,omitempty"`
+	}
+
+	// AppliedFile is part of an ApplyReport containing information about an applied file in a migration attempt.
+	AppliedFile struct {
+		migrate.File
+		Start   time.Time
+		End     time.Time
+		Skipped int      // Amount of skipped SQL statements in a partially applied file.
+		Stmts   []string // SQL statements applied with success
```
Comment (a8m) [inline cmd/atlas/internal/migrate/report.go:358]: Any reason not to call it `Applied`?

File: cmd/atlas/internal/migrate/report.go [hunk 3]
```diff
@@ -121,7 +129,7 @@ Last migration attempt had errors:
 			return buf.String(), err
 		},
 	}, ColorTemplateFuncs)
-	DefaultTemplate = template.Must(template.New("report").Funcs(StatusTemplateFuncs).Parse("{{ default . }}"))
+	StatusDefaultTemplate = template.Must(template.New("report").Funcs(StatusTemplateFuncs).Parse("{{ default . }}"))
```
Comment (a8m) [inline cmd/atlas/internal/migrate/report.go:132]: The `Default` usually comes as the first word. 
```suggestion
	DefaultStatusTemplate = template.Must(template.New("report").Funcs(StatusTemplateFuncs).Parse("{{ default . }}"))
```

**Review by a8m — COMMENTED** [review]
Looks good!

File: cmd/atlas/internal/cmdapi/migrate.go [hunk 4]
```diff
@@ -263,6 +254,13 @@ func migrateApplyRun(cmd *cobra.Command, args []string, flags migrateApplyFlags)
 			report.Log(migrate.LogDone{})
 		}
 	}
+	f := cmdmigrate.ApplyDefaultTemplate
```
Comment (a8m) [inline cmd/atlas/internal/cmdapi/migrate.go:257]:
```suggestion
	f := cmdmigrate.DefaultApplyTemplate
```

**Review by a8m — APPROVED** [review]
Last comment, can you please add a simple test case for JSON logging? Thanks

**Review by a8m — APPROVED** [review]

File: internal/integration/testdata/mysql/cli-migrate-apply.txt [hunk 5]
```diff
@@ -43,6 +43,19 @@ cmpshow pets pets.sql
...
+atlas migrate apply --url URL --revisions-schema $db --log '{{ json . }}'
+validJSON stdout
+stdout '"Driver":"mysql"'
+stdout '"Scheme":"mysql"'
+stdout '"Dir":"migrations"'
+stdout '"DSN":"root'
+stdout '"Target":"3"'
+stdout '"Pending":\[{"Name":"1_first.sql","Version":"1","Description":"first"},{"Name":"2_second.sql","Version":"2","Description":"second"},{"Name":"3_third.sql","Version":"3","Description":"third"}\]'
+stdout '"Applied":\["CREATE TABLE `users` \(`id` bigint NOT NULL AUTO_INCREMENT, `age` bigint NOT NULL, `name` varchar\(255\) NOT NULL, PRIMARY KEY \(`id`\)\) CHARSET utf8mb4 COLLATE utf8mb4_bin;"\]'
+stdout '"Start":"\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d.[0-9]"'
```
Comment (masseelch) [inline internal/integration/testdata/mysql/cli-migrate-apply.txt:57]:
```suggestion
stdout '"Start":"\d\d\d\d-\d\d-\d\dT\d\d:\d\d:\d\d.[0-9]'
```

**Pull request merged**